Add interop test cases for streaming RPCs

per-gron opened the pull request:
# server_streaming

The server_streaming test case is based on Eliza's work in #49.

When reviewing, please check if what I'm doing in `error.rs` is the idiomatic thing to do, I'm unsure. I could also do this without making the `Error` type implement `std::error::Error` but that would probably involve either messing around with refcounted assertion error lists or explicit stringing through of the errors, which would reduce readability. Another approach could be to refactor the tests to use a non-functional approach where assertions are side effectful.

I'm not sure how to best format the code for readability here. I added some non-mandatory type annotations in an attempt to make the code easier to follow, not sure what's the best practice is here

# ping_pong

The code for this test case is really not that easy to read. Besides using async/await syntax and/or fixing the error types in tower-grpc to be simpler and more unified I don't know how to write the code in a simpler way.

# Misc

Note: This will conflict with #71 so perhaps merge that first and let me resolve the conflict before reviewing this.

Fixes #49, #23.

hawkw commented on src/error.rs:
```diff
@@ -19,6 +21,66 @@ pub enum ProtocolError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match *self {
+            Error::Grpc(ref _status, ref _header_map) =>
+                write!(f, "gRPC error"),
```
Take it or leave it: When formatting a string literal with no format arguments, you could also just call `f.pad("gRPC error")` (and so on).

hawkw commented on src/error.rs:
```diff
@@ -19,6 +21,66 @@ pub enum ProtocolError {
+            Error::Grpc(ref _status, ref _header_map) =>
+                write!(f, "gRPC error"),
+            Error::Protocol(ref _protocol_error) =>
+                write!(f, "Protocol error"),
```
A bit of a nit: String descriptions of errors should typically begin with lowercase letters (as per the standard library precedent).

per-gron replied:
Ah, didn't know that. Fixed.

hawkw replied:
I think this is because the `fmt::Display` output of one error is often used as part of the message produced by the `Display` impl for another error, you'll see things like
```rust
enum MyError {
    Io(::std::io::Error),
    Proto(ProtocolError),
}
```rust
enum MyError {
    DiskIo(::std::io::Error),
    NetworkIo(::std::io::Error),
}
impl fmt::Display for MyError {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            MyError::DiskIo(ref io_err) => 
                write!(f, "error reading from disk: {}", io_err),
            MyError::NetworkIo(ref io_err) => 
                write!(f, "{} while establishing remote connection", io_err)
        }
    }
}
```
And in this case, the `Display` output for `MyErr` will look nicer if the `Display` impl for `io::Error` produces sentence fragments instead of full sentences.

This is really not terribly important in this case at all, since this isn't a library where we'd expect someone downstream to use our errors as part of their error type, but I thought I'd point it out.
```

hawkw commented on src/error.rs:
```diff
@@ -19,6 +21,66 @@ pub enum ProtocolError {
+}
+
+impl std::error::Error for ProtocolError {
+    fn description(&self) -> &str {
```
Nit: `Error::description` is deprecated now, it's suggested to leave this unimplemented and just use `fmt::Display` instead.

hawkw commented on tower-grpc-interop/src/client.rs:
```diff
@@ -278,33 +318,202 @@ impl Testcase {
+                let req = Request::new(req);
+                core.run(client
+                    .streaming_output_call(req)
+                    .map_err(|tower_err: tower_grpc::Error<tower_h2::client::Error>| -> Box<Error> {
```
I think this closure can be written as just `.map_err(Box::new)`, which is somewhat more concise.

per-gron replied:
That does not work, the compiler doesn't understand that I want this closure's return type to be `Box<Error>` (rather than `Box<tower_grpc::Error<...>>`) then.

per-gron replied:
I think this could be improved with #6. I think tower-grpc would be well served by only one error type; this is how the C++ gRPC library works.

hawkw replied:
@per-gron 
> That does not work, the compiler doesn't understand that I want this closure's return type to be `Box<Error>` (rather than `Box<tower_grpc::Error<...>>`) then.

Ah, okay. In that case, I _think_ you could write `.map_err(Box::<Error>::new)`, which is still slightly more concise --- I haven't actually checked to make sure the type checker will accept that, but I think it should...

hawkw commented on tower-grpc-interop/src/client.rs:
```diff
@@ -278,33 +318,202 @@ impl Testcase {
+                            test_assert!(
+                                "the response payload sizes should match input",
+                                RESPONSE_LENGTHS == response_lengths(&responses).as_slice(),
+                                format!("{:?}={:?}", RESPONSE_LENGTHS, response_lengths(&responses))
```
TIOLI: might be better to call `response_lengths(&responses)` once and store it in a variable...

per-gron replied:
Done

hawkw commented on tower-grpc-interop/src/client.rs:
```diff
@@ -278,33 +318,202 @@ impl Testcase {
+                /// responses from the server and assertions that were performed in
+                /// the process.
+                fn perform_ping_pong(state: State) -> ResponsesFuture {
+                    let (sender, stream, mut responses, mut assertions) =
```
Nit: I think this might be somewhat nicer as
```rust
let State { sender, stream, mut responses, mut assertions } = state;
```

per-gron replied:
Nice! I don't know why I didn't do that in the first place. Fixed.

hawkw commented on tower-grpc-interop/src/client.rs:
```diff
@@ -278,33 +318,202 @@ impl Testcase {
+                    .then(&assert_success))
+            },
+            Testcase::ping_pong => {
+                fn make_req(idx: usize) -> pb::StreamingOutputCallRequest {
```
Personally, I think it would be easier to read this code if the helper functions/types for doing the ping-pong weren't local to this function --- we're already indented several levels here. Take it or leave it.

hawkw commented on tower-grpc-interop/src/client.rs:
```diff
@@ -278,33 +318,202 @@ impl Testcase {
+                        // Convert the stream into a plain Vec
+                        response_stream.into_inner()
+                            .collect()
+                            .map_err(|tower_err: tower_grpc::Error<()>| -> Box<Error> {
```
Again, I think this can just be `map_err(Box::new)`

per-gron replied:
Same answer again. It could be made more concise with a helper function but I think a better solution would be to simplify the error types exposed by tower-grpc so that this is not needed in the first place.

hawkw commented on tower-grpc-interop/src/client.rs:
```diff
@@ -178,6 +188,34 @@ impl From<domain::resolv::error::Error> for DnsError {
+
+/// Helper function that can be used with .then to assert that the RPC performed
+/// in a test was successful.
+fn assert_success(result: Result<Vec<TestAssertion>, Box<Error>>)
```
This seems like a good idea for reducing boilerplate, thanks!

hawkw commented on tower-grpc-interop/src/client.rs:
```diff
@@ -216,6 +216,302 @@ fn assert_success(result: Result<Vec<TestAssertion>, Box<Error>>)
+    tower_h2::client::Connection<
+        tokio_core::net::TcpStream, tokio_core::reactor::Handle, tower_h2::BoxBody>>>;
+
+fn empty_unary_test(client: &mut TestClient, core: &mut tokio_core::reactor::Core)
```
Thanks, breaking these out is much nicer to read IMHO.

Take it or leave it: if you defined `TestClient` as a newtype instead,
```rust
struct TestClient(
pb::client::TestService<tower_http::AddOrigin<
    tower_h2::client::Connection<
        tokio_core::net::TcpStream, tokio_core::reactor::Handle, tower_h2::BoxBody>>>
)
```
you could then define these functions as methods:
```rust
impl TestClient {
    fn empty_unary_test(&mut self, core: &mut tokio_core::reactor::Core) -> Result<Vec<TestAssertion>, Box<Error>> {
        // ...
    }

    fn large_unary_test(&mut self, core: &mut tokio_core::reactor::Core) -> Result<...> {
        // ...
    }

    fn cacheable_unary_test(&mut self, ...) -> Result<...> {
        // ...
    }
}
```
and so on.

per-gron replied:
Done.

hawkw commented on tower-grpc-interop/src/client.rs:
```diff
@@ -216,6 +216,302 @@ fn assert_success(result: Result<Vec<TestAssertion>, Box<Error>>)
+fn empty_unary_test(client: &mut TestClient, core: &mut tokio_core::reactor::Core)
+        -> Result<Vec<TestAssertion>, Box<Error>> {
+    use pb::Empty;
+    core.run(client
```
As a side note: when I was writing this code originally, my initial intention was to have functions for each test that returned `Future<Item = Vec<TestAssertion>, Error = Box<Error>>` that we could then run on a `core` in the `main` function.

At the time, I ended up not doing that, since the returned futures weren't nameable types (as they contained closures), and it seemed like too much effort would be required to get around that. Now, I think we could do that pretty easily, with `impl Future`. 

This isn't a change you have to make in _this_ branch, but in the long run, I think it might be a nicer way to structure this code. Perhaps something that either of us could do in a follow-up branch...

per-gron replied:
That is nicer. Fixed.

hawkw commented on tower-grpc-interop/src/client.rs:
```diff
@@ -216,6 +216,302 @@ fn assert_success(result: Result<Vec<TestAssertion>, Box<Error>>)
+/// operation and performs the rest of it. It returns a future of
+/// responses from the server and assertions that were performed in
+/// the process.
+fn perform_ping_pong(state: PingPongState) -> PingPongResponsesFuture {
```
Nit: this seems like it _could_ be a method on `PingPongState`...

per-gron replied:
Done.

hawkw commented on tower-grpc-interop/src/client.rs:
```diff
@@ -216,6 +216,302 @@ fn assert_success(result: Result<Vec<TestAssertion>, Box<Error>>)
+                        }))
+                } else {
+                    sender.unbounded_send(
+                        make_ping_pong_request(responses.len())).unwrap();
```
Nit: formatting here is kind of weird, I think it would look better as:
```rust
sender.unbounded_send(
    make_ping_pong_request(responses.len())
).unwrap();
```

per-gron replied:
Heh, being used to the Google C++ code style it's your version that seems weird to me. But this isn't Google C++... Fixed.

Do you know if rustfmt is usable now? It would be nice to automatically resolve this kind of thing.

hawkw commented on tower-grpc-interop/src/client.rs:
```diff
@@ -241,281 +537,20 @@ impl Testcase {
-                    })
-                    .then(&assert_success))
-            },
+            Testcase::empty_unary =>
```
This is basically how I initially wanted to structure this function originally --- it's _so_ much less ugly than putting all the test code in this method --- but I think something got in the way of that at the time.

Thanks so much for fixing this up! 👍

hawkw commented:
@per-gron I've gone ahead and merged #71, so you'll want to rebase this branch, I think.

per-gron commented:
Rebased and addressed comments.

![Resolving my own merge conflicts](https://i.imgflip.com/2edaxq.jpg)

per-gron commented:
Ping?

hawkw commented on src/status.rs:
```diff
@@ -3,7 +3,7 @@ use std::fmt;
 use http::header::HeaderValue;
 
-#[derive(Debug, Clone)]
+#[derive(Debug, Clone, PartialEq, Eq)]
```
While adding derived traits here seems pretty harmless, I am a little concerned that it could potentially break backwards compatibility in the future, if we add fields to `Status` that can't implement `Eq` or `PartialEq`. Also, since the purpose of this branch is to add tests, I'd prefer it to not also make this change. Can we rewrite the test to use something like this?
```rust
if status.code() == Status::OK.code() {
    // ...
}
```

carllerche replied:
I agree that `Eq` and `PartialEq` should be punted to `code()`. Especially after #78 `Status` shouldn't impl these traits.

per-gron commented:
I merged from master and addressed the review comment. The merge commit has some interesting things going on in the conflicting file src/client/streaming.rs, make sure to look at that. I'm not super happy with how the readability of the logic in that function turned out.

carllerche commented:
Looks great 👍 Thanks!
